pkg/controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,226 @@
+		// if value is not none, it means this key/value need to add/update
+		if value != "" {
+			need_update = true
+			new_anno[key] = value
+		} else { // if value is empty, then this key/value need to be removed from map
+			_, exist := new_anno[key]
+			if exist {
+				delete(new_anno, key)
+				need_update = true
+			} else {
+				log.Info("No need to change ns annotation.")
+				need_update = false
+			}
+		}
+	}
+	// update k8s object
+	k8s_obj.SetAnnotations(new_anno)
+
+	// only send update request when it is needed
+	if need_update {
+		err := h.Client.Update(*ctx, k8s_obj)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *NamespaceReconciler) createVPCCR(ctx *context.Context, obj client.Object, ns string, nc_name string, vpc_name *string) (*v1alpha1.VPC, error) {
+	// check if vpc cr already exist under this namespace
+	vpcs := &v1alpha1.VPCList{}
+	r.Client.List(*ctx, vpcs, client.InNamespace(ns))
+	if len(vpcs.Items) > 0 {
+		// if there is already one vpc exist under this ns, return this vpc.
+		log.Info("vpc cr already exist, skip creating.", "VPC", vpcs.Items[0].Name)
+		return &vpcs.Items[0], nil
+	}
+	nc, nc_exist := common.ServiceMediator.GetVPCNetworkConfig(nc_name)
+	if !nc_exist {
+		message := fmt.Sprintf("missing network config %s for namespace %s", nc_name, ns)
+		r.namespaceError(ctx, obj, message, nil)
+		return nil, errors.New(message)
+	}
+
+	//TODO: in next patch, remove this validation. If user did not provide private cidr
+	// use a hardcoded cidr to create a private ip block.
+	if !common.ServiceMediator.ValidateNetworkConfig(nc) {
+		// if netwrok config is not valid, no need to retry, skip processing
+		message := fmt.Sprintf("invalid network config %s for namespace %s, missing private cidr.", nc_name, ns)
+		r.namespaceError(ctx, obj, message, nil)
+		return nil, errors.New(message)
+	}
+
+	// create vpc cr with exisitng vpc network config
+	vpc_cr := common.BuildVPCCR(ns, nc_name, vpc_name)
+	err := r.Client.Create(*ctx, vpc_cr)
+	if err != nil {
+		message := "failed to create VPC CR."
+		r.namespaceError(ctx, obj, message, err)
+		// If create VPC CR failed, put ns create event back to queue.
+		return nil, err
+	}
+
+	changes := map[string]string{
+		NS_ANNO_VPC_ERROR: "",
+	}
+	r.updateNSAnnotation(ctx, obj, changes)
+	log.Info("create vpc cr", "VPC", vpc_cr.Name, "for namespace", vpc_cr.Namespace)
+	return vpc_cr, nil
+}
+
+func (r *NamespaceReconciler) namespaceError(ctx *context.Context, k8s_obj client.Object, msg string, err error) {
+	log_err := util.If(err == nil, errors.New(msg), err).(error)
+	log.Error(log_err, msg)
+	changes := map[string]string{NS_ANNO_VPC_ERROR: msg}
+	r.updateNSAnnotation(ctx, k8s_obj, changes)
+}
+
+/*
+	VPC creation strategy:
+
+We suppose namespace should have following annotations:
+  - "nsx.vmware.com/vpc_name": "kube-system/<Supervisor ID>"
+    If the ns contains this annotation, first check if	the namespace in annotation is the same as
```

Comment:
`the namespace in annotation` means `kube-system` in L134? If yes, I prefer to change `"nsx.vmware.com/vpc_name": "kube-system/<Supervisor ID>"` to `"nsx.vmware.com/vpc_name": "<namespace name>/<Supervisor ID>"`